File: apps/api/schemas/uploads.py
from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class InitializeUploadRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    filename: str = Field(min_length=1, max_length=512)
    media_type: str = Field(min_length=1, max_length=255)
    expected_size: int = Field(gt=0)
    expected_sha256: str = Field(pattern=r"^[a-f0-9]{64}$")
    part_size: int = Field(default=8 * 1024 * 1024, ge=1024, le=64 * 1024 * 1024)


class UploadResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    filename: str
    media_type: str
    expected_size: int
    expected_sha256: str
    part_size: int
    status: str
    completed_asset_id: UUID | None
    error_code: str | None


class UploadPartResponse(BaseModel):
    upload_id: UUID
    part_number: int
    byte_size: int
    sha256: str
    duplicate: bool


class CompleteUploadResponse(BaseModel):
    upload_id: UUID
    source_video_id: UUID
    asset_id: UUID
    sha256: str
    byte_size: int
    status: str


class DownloadURLResponse(BaseModel):
    asset_id: UUID
    url: str
    expires_in_seconds: int

File: apps/web/src/api/uploads.ts
import { apiClient, type ApiResponse, type VidGenClient } from "./client";

export interface UploadSession {
  id: string;
  project_id: string;
  filename: string;
  media_type: string;
  expected_size: number;
  expected_sha256: string;
  part_size: number;
  status: string;
  completed_asset_id: string | null;
  error_code: string | null;
}

export interface UploadPartResult {
  upload_id: string;
  part_number: number;
  byte_size: number;
  sha256: string;
  duplicate: boolean;
}

export interface CompletedUpload {
  upload_id: string;
  source_video_id: string;
  asset_id: string;
  sha256: string;
  byte_size: number;
  status: string;
}

export interface InitializeUploadInput {
  filename: string;
  media_type: string;
  expected_size: number;
  expected_sha256: string;
  part_size: number;
}

export function initializeUpload(
  projectId: string,
  input: InitializeUploadInput,
  client: VidGenClient = apiClient,
): Promise<ApiResponse<UploadSession>> {
  return client.post<UploadSession>(`/api/v1/projects/${projectId}/uploads`, { body: input });
}

export function uploadPart(
  uploadId: string,
  partNumber: number,
  bytes: ArrayBuffer | Uint8Array | Blob,
  client: VidGenClient = apiClient,
  signal?: AbortSignal,
): Promise<ApiResponse<UploadPartResult>> {
  return client.put<UploadPartResult>(`/api/v1/uploads/${uploadId}/parts/${partNumber}`, {
    rawBody: bytes as BodyInit,
    contentType: "application/octet-stream",
    ...(signal ? { signal } : {}),
  });
}

export function completeUpload(
  uploadId: string,
  client: VidGenClient = apiClient,
): Promise<ApiResponse<CompletedUpload>> {
  return client.post<CompletedUpload>(`/api/v1/uploads/${uploadId}/complete`);
}

export interface DownloadUrl {
  asset_id: string;
  url: string;
  expires_in_seconds: number;
}

/**
 * Signed download URLs are requested shortly before use and never persisted:
 * they are ephemeral capabilities, not application state.
 */
export function getDownloadUrl(
  assetId: string,
  client: VidGenClient = apiClient,
  signal?: AbortSignal,
): Promise<ApiResponse<DownloadUrl>> {
  return client.get<DownloadUrl>(
    `/api/v1/assets/${assetId}/download-url`,
    signal ? { signal } : {},
  );
}
